tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestApalacheBuilder.scala

```diff
@@ -250,10 +251,286 @@ class TestApalacheBuilder extends BuilderTest {
     }
   }
 
-  // TODO: Implement builder.letInEx, so we can construct lambdas for the following tests
-  test("mkSeq") {}
-  test("foldSet") {}
-  test("foldSeq") {}
+  object LambdaFactory {
+    // we pass body and bodyT separately, so we can avoid building `body` to get its type.
+    // when using Gen[TParam] we get the body type explicitly anyway.
+    // Assumption: build(body).typeTag.asTlaType1() == bodyT
+    /**
+     * Creates a lambda expression of the form
+     * {{{
+     * LET opName(p1,...,pn) == body IN opName }}} ``
+     */
+    def mkLambda(
+        opName: String,
+        tparams: Seq[TypedParam],
+        body: TBuilderInstruction,
+        bodyT: TlaType1): TBuilderInstruction = {
+      val paramTs = tparams.map { _._2 }
+      val operT = OperT1(paramTs, bodyT)
+      builder.letIn(
+          builder.name(opName, operT),
+          builder.decl(opName, body, tparams: _*),
+      )
+    }
+  }
+
+  test("mkSeq") {
+    import LambdaFactory.mkLambda
+    type T = (TBuilderInstruction, TBuilderInstruction)
+    type TParam = (Int, TlaType1)
+
+    implicit val gen: Gen[TParam] = Gen.zip(Gen.choose(0, 10), singleTypeGen)
+
+    // MkSeq(n, LET F(i) == e IN F)
+    def mkWellTyped(tparam: TParam): T = {
+      val (n, t) = tparam
+      val param = builder.param("i", IntT1)
+      (
+          builder.int(n),
+          mkLambda(
+              "F",
+              Seq(param),
+              builder.name("e", t),
+              t,
+          ),
```

Comment:
why is it called lambda, if it has a name (that is, "F")? I thought, lambdas were anonymous functions/operators.

Reply:
It's the name of the form `LET F(...) == ... IN F`. actual TLA `LAMBDA` expressions get translated into this, but it's also the form needed in, for example, `FoldSet`.